bpf/bpf_lxc.c

```diff
@@ -2372,21 +2427,74 @@ int cil_to_container(struct __ctx_buff *ctx)
 __declare_tail(CILIUM_CALL_IPV4_POLICY_DENIED)
 int tail_policy_denied_ipv4(struct __ctx_buff *ctx)
 {
+	void *data, *data_end;
+	struct iphdr *ip4;
+	struct ct_state ct_state_local = {};
+	#if defined(TUNNEL_MODE)
+    ct_state_local.from_tunnel = ctx_load_and_clear_meta(ctx, CB_FROM_TUNNEL) ? 1 : 0;
+    #endif
+	struct remote_endpoint_info *info;
+	struct trace_ctx trace = {
+		.reason = TRACE_REASON_POLICY,
+		.monitor = TRACE_PAYLOAD_LEN,
+	};
 	int ret;
+	__s8 ext_err = 0;
 	__u32 verdict = ctx_load_meta(ctx, CB_VERDICT);
+	__u8 metric_dir = (__u8)ctx_load_meta(ctx, CB_POLICY_DENY_DIR);
+	__u32 src_label = metric_dir == METRIC_EGRESS ?
+			  SECLABEL_IPV4 : ctx_load_meta(ctx, CB_POLICY_DENY_SRC_ID);
+	__u32 dst_sec_identity = WORLD_IPV4_ID;
 
 	ret = generate_icmp4_reply(ctx, ICMP_DEST_UNREACH, ICMP_PKT_FILTERED);
 	if (!ret) {
+		if (!revalidate_data(ctx, &data, &data_end, &ip4)) {
+			ret = DROP_INVALID;
+			goto drop_err;
+		}
+
 		cilium_dbg_capture(ctx, DBG_CAPTURE_DELIVERY, ctx_get_ifindex(ctx));
-		ret = redirect_self(ctx);
 
-		if (!IS_ERR(ret)) {
-			update_metrics(ctx_full_len(ctx), METRIC_EGRESS, __DROP_REASON(verdict));
+		info = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		if (info)
+			dst_sec_identity = info->sec_identity;
+
+		/* Forward the ICMP packet to its destination following the same
+		 * path as regular egress traffic from the pod would take.
+		 */
+		ret = ipv4_forward_to_destination(ctx, ip4, NULL, &dst_sec_identity,
+						  &ext_err, &ct_state_local, CT_NEW,
+						  info, &trace, false, false, false, 0, 0);
```

Comment:
Oh it's going to be fun figuring out all the little corner-cases from passing "default" values here :)